GCS blob store: add OperationPurpose/Operation stats counters

mhl-b opened the pull request:
Add counters for operations and requests for each combination of OperationPurpose and Operation for GCS blob store. Align with other BlobStore implementations (AWS and Azure).

Refactored `GoogleCloudStorageHttpStatsCollector`. Removed indirection of `trackers` and `trackerFactories`.

ES-10213
ES-10863

nicktindall commented on modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageOperationsStats.java:
```diff
@@ -10,53 +10,75 @@
-    void trackPutOperation() {
-        putCount.incrementAndGet();
+    void trackOperation(OperationPurpose purpose, Operation operation) {
+        operationCounters(purpose, operation)[0].add(1);
```
Nit: maybe have constants for the `0` and `1` to make it clear what they mean?

nicktindall replied:
in fact, instead of having a two dimensional array, why don't we have something like an `OperationTracker` which has `trackOperation(OperationPurpose operation)` and `trackRequest(OperationPurpose operation)` methods?

Either have an array of those indexed by `Operation` ordinal, they can each contain an array of `LongAdder` keyed by `OperationPurpose` ordinal. I think it might make it clearer and allow us to have those objects host the logic for writing the APM metrics.

mhl-b replied:
addressed https://github.com/elastic/elasticsearch/pull/122991/commits/b60fcc83400fece92656d396f4a757ce901098e1

nicktindall commented on modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageOperationsStats.java:
```diff
@@ -10,53 +10,75 @@
+        Operation(String name) {
+            this.name = name;
+        }
+    }
```
Also `Post` and `Put` are not entirely clear. Both types of upload begin with a `POST`, the resumable one uses `PUT` for the chunks.

nicktindall replied:
Agreed we'll raise this with relevant people and address later.

mhl-b committed:
wip

mhl-b committed:
tidy

nicktindall commented on modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:
```diff
@@ -380,7 +386,7 @@ public void write(byte[] b, int off, int len) throws IOException {
                                 while (written < len) {
                                     // at most write the default chunk size in one go to prevent allocating huge buffers in the SDK
                                     // see com.google.cloud.BaseWriteChannel#DEFAULT_CHUNK_SIZE
```
Nit: we can probably remove the `see ...` comment because it's proper javadoc above now

nicktindall commented on modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageHttpStatsCollector.java:
```diff
@@ -9,115 +9,82 @@
+                // ignore other post requests
+            }
+            default -> {
+                // ignore other http methods
```
with all the methods/paths/params that we ignore, have we verified that we don't use them? might me worth logging at debug or something just to see if there's something we missed?

nicktindall reviewed (approved):
LGTM!

mhl-b committed:
Merge remote-tracking branch 'upstream/main' into gcp-metering

mhl-b committed:
Merge remote-tracking branch 'upstream/main' into gcp-metering

mhl-b committed:
tests and feedback

mhl-b committed:
Update docs/changelog/122991.yaml

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

elasticsearchmachine commented:
Hi @mhl-b, I've created a changelog YAML for you.

mhl-b commented:
@nicktindall , please take a look at the last commit https://github.com/elastic/elasticsearch/pull/122991/commits/a5549e4d93a023f8d827e4fdfecac4b8efdfdad2

I updated CAS test, added assertion for ignored request, and minor refactor in the test code.

nicktindall commented on modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobContainerStatsTests.java:
```diff
@@ -236,4 +272,48 @@ protected String getEndpointForServer(final HttpServer server) {
+        }
+
+        StatsMap add(Operation operation, long ops, long reqs) {
+            return add(purpose, operation, ops, reqs);
```
Do we need to assert that `this.purpose` isn't null if this is called?

mhl-b replied:
I can remove nullability completely, there is no place that init with null.

mhl-b replied:
https://github.com/elastic/elasticsearch/pull/122991/commits/4a975287c47d05a38a3fd06ed1f9eeed34096457

nicktindall reviewed (commented):
LGTM

mhl-b committed:
test update

mhl-b committed:
Merge branch 'gcp-metering' of github.com:mhl-b/elasticsearch into gcp-metering

mhl-b committed:
change operations and more test fixes

nicktindall commented on modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageHttpStatsCollector.java:
```diff
@@ -9,115 +9,75 @@
-            this.statsTracker = statsTracker;
+        var respCode = response.getStatusCode();
+        // Some of the intermediate and error codes are still counted as "good" requests
+        if (((respCode >= 200 && respCode < 300) || respCode == 308 || respCode == 404) == false) {
```
Nit: for the first condition we could use `org.elasticsearch.rest.RestStatus#isSuccessful`, that logic was repeated in lots of places. Unless you think that obscures it.

mhl-b replied:
I think it's more readable in this form since multiple codes are handled. I'm more concerned about codes I dont handle here, but we would like to.

nicktindall commented on modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageOperationsStats.java:
```diff
@@ -10,53 +10,166 @@
+            this.tracker = new ServerlessTracker(bucketName);
+        } else {
+            this.tracker = new StatefulTracker();
+        }
```
I like!

nicktindall reviewed (approved):
Very nice

mhl-b committed:
change assert to logging

mhl-b committed:
Merge remote-tracking branch 'upstream/main' into gcp-metering

mhl-b committed:
Merge branch 'main' into gcp-metering

mhl-b committed:
Merge remote-tracking branch 'upstream/main' into gcp-metering

mhl-b committed:
Merge branch 'gcp-metering' of github.com:mhl-b/elasticsearch into gcp-metering
